src/main/java/Duke.java

```diff
@@ -1,10 +1,167 @@
+    private Storage storage;
+    private TaskList tasks;
+    private ScrollPane scrollPane;
+    private VBox dialogContainer;
+    private TextField userInput;
+    private Button sendButton;
+    private Scene scene;
+
+    private Image user = new Image(this.getClass().getResourceAsStream("/images/DaUser.png"));
+    private Image duke = new Image(this.getClass().getResourceAsStream("/images/DaDuke.png"));
+    // Empty constructor required for Launcher
+    public Duke() {
+    }
+
+    public Duke(String filePath) {
+        storage = new Storage(filePath);
+        try {
+            tasks = new TaskList(storage.loadTasksFromSaveFile());
+
+            // Since we can successfully load it, we save all in TaskList
+            storage.writeTasks(tasks);
+        } catch (FileNotFoundException e) {
+            // Load an empty one
+            tasks = new TaskList();
+            Ui.printLine();
+            Ui.indent("It seems that there is no save file in: " + filePath);
+            Ui.printLine();
+        }
+    }
+
+    public void run() {
+        Ui.greet();
+        while (Parser.stillHasInputs()) {
+            try {
+                String input = Parser.readLine();
+                if (Parser.commandEquals("bye", input)) {
+                    Ui.sayGoodbye();
+                    break;
+                } else if (Parser.commandEquals("list", input)) {
+                    Ui.displayList(tasks);
+                } else if (Parser.commandEquals("done", input)) {
+                    int taskNo = Parser.getTaskNo(input);
+                    tasks.markTaskAsDone(taskNo);
+                } else if (Parser.commandEquals("delete", input)) {
+                    int taskNo = Parser.getTaskNo(input);
+                    tasks.deleteTask(taskNo);
+                } else if (Parser.commandEquals("deadline", input)) {
+                    if (Parser.hasNoArgs(input)) throw new EmptyDescriptionException();
+                    if (tasks.isFull()) throw new TooManyTasksException();
+
+                    tasks.addDeadline(Parser.getArgs(input));
+                } else if (Parser.commandEquals("event", input)) {
+                    if (Parser.hasNoArgs(input)) throw new EmptyDescriptionException();
+                    if (tasks.isFull()) throw new TooManyTasksException();
+
+                    tasks.addEvent(Parser.getArgs(input));
+                } else if (Parser.commandEquals("todo", input)) {
+                    if (Parser.hasNoArgs(input)) throw new EmptyDescriptionException();
+                    if (tasks.isFull()) throw new TooManyTasksException();
+
+                    tasks.addTodo(Parser.getArgs(input));
+                } else if (Parser.commandEquals("find", input)) {
+                    if (Parser.hasNoArgs(input)) throw new EmptyDescriptionException();
+
+                    tasks.find(Parser.getArgs(input));
+                } else {
+                    throw new UnknownCommandException();
+                }
```

Comment:
good job with the function and variable namings! it makes it very easy for me (and any other developers) to read through! one improvement that you could consider is to abstract this to another class called Processor or something, so that your main class looks shorter.